### PR TITLE
[cpp] Write custom CharStream to handle continuation characters

### LINKED ISSUE #448: [cpp] Write custom CharStream to handle continuation characters
For reference, check https://github.com/pmd/pmd/pull/446#issuecomment-309231096

### PR DESCRIPTION
**Pull request by adangel**
Fixes #448 

Before submitting a PR, please check that:
 - [x] The PR is submitted against `master`. The PMD team will merge back to support branches as needed.
 - [x] `mvn test` passes.
 - [x] `mvn checkstyle:check` passes. [Check this for more info](https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md#code-style)

### CODE REVIEW

File: pmd-cpp/src/test/java/net/sourceforge/pmd/cpd/CPPTokenizerContinuationTest.java [hunk 1]
```diff
@@ -25,35 +24,32 @@
...
 
-        // Note: the token should be "ab", but since we skip "\\\n" between a and b,
-        // we end up with two tokens. -> the tokens are not joined.
-        // This could lead to false negatives in duplication detection.
-        // However, this is only a problem, if the continuation character is added *within*
-        // a token and not at token boundaries.
+        // special case, if the continuation is *within* a token
         // see also test #testContinuationIntraToken
-        //assertEquals("ab", findByLine(8, tokens).get(2).toString());
+        assertEquals("ab", findByLine(8, tokens).get(2).toString());
```
Comment (jsotuyod) [inline pmd-cpp/src/test/java/net/sourceforge/pmd/cpd/CPPTokenizerContinuationTest.java]: maybe add an assert checking the token's beingLine / endLine are properly filled?

Comment (adangel) [inline pmd-cpp/src/test/java/net/sourceforge/pmd/cpd/CPPTokenizerContinuationTest.java]: Good idea, I'll add this.
About the grammar, I just remembered this, too :)

**Comment by adangel** [comment]
Done 😄

**Comment by jsotuyod** [comment]
Amazing! Thanks!

**Pull request merged**